### ISSUE TITLE
Rewrite the event system

### ISSUE THREAD

**Issue by infogulch** [infogulch/pyspades#287]
(There's a TL;DR at the end, fyi)
## Problems

The current event handling system is less than ideal. There are several problems with the "inherit to infinity" model:
- Ancestor classes cannot raise an event from their descendant. I.e. the order that scripts are loaded matters.
- All methods _must_ call the base class's method to chain an event. It's way too easy to break the chain, and not obvious for new scripters.
- Ancestor classes cannot block an event from the descendants without the descendants doing some base class method-calling hackery.
- If an event handler is malformed (e.g. wrong number of arguments) or throws an unexpected exception the event is cancelled and breaks pyspades.
- Event handlers cannot be unregistered. (i.e. classes can't be un-inherited) This prevents modules from being unloaded or reloaded on the fly.
- Raising an event that has no handlers throws an error about the method not being defined.
- Each event must traverse the entire inherited hierarchy of classes to check if any of them handle the event, which is inefficient.
## Required Features

So Pyspades needs a new event system. Features that the new system should have:
- Remove (or at least significantly reduce) the effect that the order in which scripts are loaded has on their functionality.
- Make it harder to accidentally break an event, and reduce the requirements of handling an event.
- Print errors to the console when a malformed or error-prone handler is called instead of crapping out.
- Any script should be able to block, receive, or raise any event, regardless of whether that event is handled or raised in any other script.
- Be able to remove all event handlers that a script made at any time.
- Be able to add new event handlers at any time.
### Handling an event

Scripts can register event handlers by name at one of three different levels, `BLOCK`, `CONSUME`, and `NOTIFY`. When an event is raised the registered handlers in each level are called in level-order.
- `BLOCK`: Event handlers registered in this level are called first. If a handler returns a result that evaluates to `true`, then the event is cancelled, no further handlers are called, and the result is returned to the event raiser. This level is meant for **permissions**: return true if _nothing should happen at this event at all_, and avoid putting handlers that have side effects in this level.
- `CONSUME`: Handlers in this level are called next. Handlers should assume the event is already valid (since it was not `BLOCK`ed). This level is meant for event **consumers**: handlers that stop the normal procedure event _to do something else instead_. Handlers should return true after they finished their alternative procedure, subsequent handlers are not called, and the result is returned to the event raiser.
- `NOTIFY`: Default level. Handlers in this level are called last. The "normal" procedure for an event should register a handler here. Return values from registered handlers in this level are ignored, and all handlers in this level are called once handlers in the previous levels are finished. (I.e. either _all_ handlers in this level are called or _none_ of them are.) When all handlers in this level are finished `None` is returned to the event raiser. Use this level for the normal procedure of an event, for handlers that want to do something _in addition to_ the normal procedure, or for handlers that want to track successful events (stat counters etc.).
### Raising an event

All scripts can raise events with parameters and receive the result.
## TL;DR

The current event system sucks and is missing many features.

I have a proposal that I believe fixes all the problems and covers all the required features.

This has been floating around in my head for the last few months, and I've given a fair deal of thought to it. However I haven't gotten many 3rd party opinions on it. If you have any questions or additions for anything I would greatly appreciate it, as a fresh brain and second opinion can help find issues early on which is important at the high-level design phase this feature is in at the moment.

This issue is missing example use-cases and fixes. If you would like me to cover something specific, just ask.

**Comment by infogulch** [comment]
So the new example implementation is too big for a gist (gists don't like directories :frowning:) so I made a temporary github repo for it: https://github.com/infogulch/pyspades-events
